File: backend/app/schemas/order.py
import uuid
from datetime import datetime
from decimal import Decimal

from app.models.enums import DeliveryMethod, DocumentType, OrderStatus, PaymentMethod, PurchaseType
from app.schemas.address import AddressIn
from pydantic import BaseModel, ConfigDict, model_validator


class BillingDocumentIn(BaseModel):
    document_type: DocumentType

    # COMPANY_INVOICE fields (manual entry for PRIVATE mode; ignored in COMPANY mode)
    company_name: str | None = None
    company_nip: str | None = None

    # PERSONAL_INVOICE fields
    first_name: str | None = None
    last_name: str | None = None

    # Billing address (RECEIPT → empty; PERSONAL/COMPANY invoice → required in PRIVATE mode)
    billing_street: str | None = None
    billing_city: str | None = None
    billing_postal_code: str | None = None
    billing_country: str | None = None

    @model_validator(mode="after")
    def validate_consistency(self) -> "BillingDocumentIn":
        if self.document_type == DocumentType.COMPANY_INVOICE:
            has_name = bool(self.company_name)
            has_nip = bool(self.company_nip)
            if has_name != has_nip:
                raise ValueError("company_name and company_nip must be provided together")
        if self.document_type == DocumentType.PERSONAL_INVOICE:
            has_first = bool(self.first_name)
            has_last = bool(self.last_name)
            if has_first != has_last:
                raise ValueError("first_name and last_name must be provided together")
        return self


class BillingDocumentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    document_type: DocumentType
    document_number: str | None
    company_name: str | None
    company_nip: str | None
    first_name: str | None
    last_name: str | None
    billing_street: str | None
    billing_city: str | None
    billing_postal_code: str | None
    billing_country: str | None
    pdf_url: str | None
    issued_at: datetime | None
    created_at: datetime


class OrderCreate(BaseModel):
    product_ids: list[uuid.UUID]
    purchase_type: PurchaseType
    document: BillingDocumentIn
    address_id: uuid.UUID | None = None
    shipping_address: AddressIn | None = None
    save_address: bool = False

    delivery_method: DeliveryMethod
    payment_method: PaymentMethod
    recipient_name: str
    recipient_phone: str
    recipient_email: str | None = None
    note: str | None = None


class ShipmentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    delivery_method: DeliveryMethod
    shipping_cost: Decimal
    recipient_name: str
    recipient_phone: str
    recipient_email: str | None
    shipping_street: str
    shipping_city: str
    shipping_postal_code: str
    shipping_country: str
    created_at: datetime


class OrderItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    product_id: uuid.UUID
    product_name: str
    product_sku: str
    quantity: int
    unit_price: Decimal
    discount_percentage: Decimal


class OrderOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    status: OrderStatus
    purchase_type: PurchaseType
    company_id: uuid.UUID | None = None
    payment_method: PaymentMethod
    total_amount: Decimal
    note: str | None
    created_at: datetime

    billing_document: BillingDocumentOut
    shipment: ShipmentOut

    items: list[OrderItemOut]


class OrderSummaryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    status: OrderStatus
    purchase_type: PurchaseType
    company_id: uuid.UUID | None = None
    total_amount: Decimal
    created_at: datetime
    item_count: int


class DeliveryOptionOut(BaseModel):
    delivery_method: DeliveryMethod
    cost: Decimal


class PaymentOptionOut(BaseModel):
    payment_method: PaymentMethod
    b2b_only: bool


class CheckoutOptionsOut(BaseModel):
    delivery_methods: list[DeliveryOptionOut]
    payment_methods: list[PaymentOptionOut]


class MockPaymentIn(BaseModel):
    success: bool

File: frontend/src/api/order.ts
import apiFetch from "./client";
import type { AddressIn } from "./address";

export type PurchaseType = "B2B" | "B2C";
export type DocumentType = "RECEIPT" | "PERSONAL_INVOICE" | "COMPANY_INVOICE";
export type OrderStatus =
  | "PENDING_PAYMENT"
  | "PREPARING"
  | "SHIPPED"
  | "DELIVERED"
  | "CANCELLED"
  | "RETURNED";
export type DeliveryMethod =
  | "COURIER"
  | "COURIER_EXPRESS"
  | "INPOST_LOCKER"
  | "PICKUP";
export type PaymentMethod =
  | "BANK_TRANSFER"
  | "CARD"
  | "BLIK"
  | "CASH_ON_DELIVERY"
  | "DEFERRED";

export interface BillingDocumentIn {
  document_type: DocumentType;
  company_name?: string;
  company_nip?: string;
  first_name?: string;
  last_name?: string;
  billing_street?: string;
  billing_city?: string;
  billing_postal_code?: string;
  billing_country?: string;
}

export interface BillingDocumentOut {
  id: string;
  document_type: DocumentType;
  document_number: string | null;
  company_name: string | null;
  company_nip: string | null;
  first_name: string | null;
  last_name: string | null;
  billing_street: string | null;
  billing_city: string | null;
  billing_postal_code: string | null;
  billing_country: string | null;
  pdf_url: string | null;
  issued_at: string | null;
  created_at: string;
}

export interface ShipmentOut {
  delivery_method: DeliveryMethod;
  shipping_cost: string;
  recipient_name: string;
  recipient_phone: string;
  recipient_email: string | null;
  shipping_street: string;
  shipping_city: string;
  shipping_postal_code: string;
  shipping_country: string;
  created_at: string;
}

export interface OrderItemOut {
  id: string;
  product_id: string;
  product_name: string;
  product_sku: string;
  product_slug: string | null;
  quantity: number;
  unit_price: string;
  discount_percentage: string;
}

export interface OrderOut {
  id: string;
  status: OrderStatus;
  purchase_type: PurchaseType;
  company_id: string | null;
  payment_method: PaymentMethod;
  total_amount: string;
  note: string | null;
  created_at: string;
  billing_document: BillingDocumentOut;
  shipment: ShipmentOut;
  items: OrderItemOut[];
}

export interface OrderSummaryOut {
  id: string;
  status: OrderStatus;
  purchase_type: PurchaseType;
  company_id: string | null;
  total_amount: string;
  created_at: string;
  item_count: number;
}

export interface OrderCreate {
  product_ids: string[];
  purchase_type: PurchaseType;
  document: BillingDocumentIn;
  address_id?: string;
  shipping_address?: AddressIn;
  save_address?: boolean;
  delivery_method: DeliveryMethod;
  payment_method: PaymentMethod;
  recipient_name: string;
  recipient_phone: string;
  recipient_email?: string;
  note?: string;
}

export interface DeliveryOptionOut {
  delivery_method: DeliveryMethod;
  cost: string;
}

export interface PaymentOptionOut {
  payment_method: PaymentMethod;
  b2b_only: boolean;
}

export interface CheckoutOptionsOut {
  delivery_methods: DeliveryOptionOut[];
  payment_methods: PaymentOptionOut[];
}

export const DELIVERY_LABELS: Record<DeliveryMethod, string> = {
  COURIER: "Courier",
  COURIER_EXPRESS: "Courier (Express)",
  INPOST_LOCKER: "InPost Locker",
  PICKUP: "Personal pickup",
};

export const PAYMENT_LABELS: Record<PaymentMethod, string> = {
  BANK_TRANSFER: "Bank transfer",
  CARD: "Credit / debit card",
  BLIK: "BLIK",
  CASH_ON_DELIVERY: "Cash on delivery",
  DEFERRED: "Deferred payment (invoice)",
};

export const ORDER_STATUS_LABELS: Record<OrderStatus, string> = {
  PENDING_PAYMENT: "Awaiting payment",
  PREPARING: "Preparing",
  SHIPPED: "Shipped",
  DELIVERED: "Delivered",
  CANCELLED: "Cancelled",
  RETURNED: "Returned",
};

export const createOrder = (data: OrderCreate): Promise<OrderOut> =>
  apiFetch<OrderOut>("/orders", { method: "POST", body: data });

export const getOrders = (purchaseType?: PurchaseType): Promise<OrderSummaryOut[]> => {
  const qs = purchaseType ? `?purchase_type=${purchaseType}` : "";
  return apiFetch<OrderSummaryOut[]>(`/orders${qs}`);
};

export const getOrder = (id: string): Promise<OrderOut> =>
  apiFetch<OrderOut>(`/orders/${id}`);

export const getCheckoutOptions = (): Promise<CheckoutOptionsOut> =>
  apiFetch<CheckoutOptionsOut>("/orders/checkout-options");

export const mockPayment = (orderId: string, success: boolean): Promise<OrderOut> =>
  apiFetch<OrderOut>(`/orders/${orderId}/payment/mock`, {
    method: "POST",
    body: { success },
  });

export const confirmPayment = (orderId: string): Promise<OrderOut> =>
  apiFetch<OrderOut>(`/orders/${orderId}/payment/confirm`, { method: "POST" });

export const advanceOrderStatus = (orderId: string): Promise<OrderOut> =>
  apiFetch<OrderOut>(`/orders/${orderId}/advance-status`, { method: "POST" });
